### PR TITLE
Sync from `rust-lang/rust`

### PR DESCRIPTION
**Pull request by fasterthanlime**
This is the first subtree sync PR from `rust-lang/rust` to `rust-lang/rust-analyzer`, made by following the instructions at: https://doc.rust-lang.org/nightly/clippy/development/infrastructure/sync.html#performing-the-sync-from-rust-langrust-to-clippy

It includes all the commits that were made in:

  * https://github.com/rust-lang/rust/pull/99603

(And that were reviewed ahead of time by @Veykril + @lnicola)

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,7 +1,8 @@
 * text=auto eol=lf
+
 # git grep shouldn't match entries in this benchmark data
 bench_data/** binary
-crates/syntax/test_data/** -text eof=LF
```
Comment (bjorn3) [inline .gitattributes:4]: Why was this removed?

File: crates/hir-def/src/import_map.rs [hunk 2]
```diff
@@ -167,11 +167,7 @@ fn collect_import_map(db: &dyn DefDatabase, krate: CrateId) -> ImportMap {
 
         let visible_items = mod_data.scope.entries().filter_map(|(name, per_ns)| {
             let per_ns = per_ns.filter_visibility(|vis| vis == Visibility::Public);
-            if per_ns.is_none() {
-                None
-            } else {
-                Some((name, per_ns))
-            }
+            if per_ns.is_none() { None } else { Some((name, per_ns)) }
```
Comment (bjorn3) [inline crates/hir-def/src/import_map.rs:170]: This is the result of some nightly only options rust-lang/rust uses: https://github.com/rust-lang/rust/blob/2f320a224e827b400be25966755a621779f797cc/rustfmt.toml#L2-L4 Stable rustfmt without those options as used in this repo will revert these changes. You will need to do one of:

* Copy the options to a rustfmt.toml in this repo and switch to nightly rustfmt
* Ignore rust-analyzer for rustfmt in rust-lang/rust
* Explicitly use the stable defaults of those options in rustfmt.toml in this repo and hope rustfmt will pick them in favor of the global rustfmt.toml of rust-lang/rust

File: .gitattributes (hunk 1 above)
Comment (fasterthanlime) [reply .gitattributes:4]: It's a fun story! The short answer it: it caused bors to fail to merge the `rust-lang/rust` subtree PR.

The longer story is: that change was introduced in https://github.com/rust-lang/rust-analyzer/pull/2166 - but it never did what was intended.

`eof` is not a gitattribute setting. `eol` is, but even that line used `eol`, it _still_ wouldn't do anything, because of `-text`. So by mistake, 3 files were treated as binary, and checked out with CRLF line endings (those were the ones bors choked on).

After removing this line from `.gitattribute`, I re-added the folder to git, removed it from disk and used `git restore` to write it with the proper line endings. The relevant commit is here: https://github.com/rust-lang/rust/commit/1bb63df1fdd40cbb0f3b1c9d5a838f9195142a07

File: crates/syntax/test_data/parser/validation/0045_ambiguous_trait_object.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
-type Foo = fn() -> dyn Send + 'static;
-fn main() {
-    let b = (&a) as &dyn Add<Other, Output = Addable> + Other;
-}
+type Foo<'a> = &'a dyn Send + Sync;
+type Foo = *const dyn Send + Sync;
+type Foo = fn() -> dyn Send + 'static;
+fn main() {
+    let b = (&a) as &dyn Add<Other, Output = Addable> + Other;
+}
```
Comment (bjorn3) [inline crates/syntax/test_data/parser/validation/0045_ambiguous_trait_object.rs:6]: What is the difference in this file?

File: crates/hir-def/src/import_map.rs (hunk 2 above)
Comment (fasterthanlime) [reply crates/hir-def/src/import_map.rs:170]: That one's my bad, I had "format on save" set up when doing those edits. I'll revert them in this PR.

**Comment by bjorn3** [comment]
Please make sure to fix my review comments in additional commits to avoid confusing git subtree.

File: crates/syntax/test_data/parser/validation/0045_ambiguous_trait_object.rs (hunk 3 above)
Comment (Veykril) [reply crates/syntax/test_data/parser/validation/0045_ambiguous_trait_object.rs:6]: Line endings I would assume?

Comment (fasterthanlime) [reply crates/syntax/test_data/parser/validation/0045_ambiguous_trait_object.rs:6]: CR => LF (as for the other 2 files in the same folder), see https://github.com/rust-lang/rust-analyzer/pull/12871#discussion_r928817419

**Comment by fasterthanlime** [comment]
> Please make sure to fix my review comments in additional commits to avoid confusing git subtree.

@bjorn3 Do you mean avoiding to squash/rebase this PR? If I `git commit` on that branch and push to `fasterthanlime:sync-from-rust-2` we should be fine, right?

File: .gitattributes (hunk 1 above)
Comment (fasterthanlime) [reply .gitattributes:4]: I just did a fresh clone of that branch from Windows and ran `cargo test` with the latest stable, no issues whatsoever.

I'm still convinced #2166 wasn't needed, as explained here: https://github.com/rust-lang/rust-analyzer/pull/2166#discussion_r928246345

**Comment by bjorn3** [comment]
> Do you mean avoiding to squash/rebase this PR?

Indeed

> If I git commit on that branch and push to fasterthanlime:sync-from-rust-2 we should be fine, right?

Yes

**Review by bjorn3 — APPROVED** [review]

**Review by Veykril — APPROVED** [review]
